File: ios/new/HybridRiveView.swift
@_spi(RiveExperimental) import RiveRuntime
import Foundation
import NitroModules
import UIKit

private struct DefaultConfiguration {
  static let autoPlay = true
}

typealias HybridDataBindMode = Variant__any_HybridViewModelInstanceSpec__DataBindMode_DataBindByName

extension Optional
where Wrapped == HybridDataBindMode {
  func toExperimentalBindData() throws -> ExperimentalBindData {
    guard let value = self else {
      return .auto
    }

    switch value {
    case .first(let viewModelInstance):
      if let instance = (viewModelInstance as? HybridViewModelInstance)?.viewModelInstance {
        return .instance(instance)
      } else {
        throw RuntimeError.error(withMessage: "Invalid ViewModelInstance")
      }
    case .second(let mode):
      switch mode {
      case .auto:
        return .auto
      case .none:
        return .none
      }
    case .third(let dataBindByName):
      return .byName(dataBindByName.byName)
    }
  }

  func isEqual(to other: HybridDataBindMode?) -> Bool {
    guard let lhs = self, let rhs = other else {
      return self == nil && other == nil
    }

    switch (lhs, rhs) {
    case (.first(let lhsInstance), .first(let rhsInstance)):
      let lhsVMI = (lhsInstance as? HybridViewModelInstance)?.viewModelInstance
      let rhsVMI = (rhsInstance as? HybridViewModelInstance)?.viewModelInstance
      return lhsVMI === rhsVMI
    case (.second(let lhsMode), .second(let rhsMode)):
      return lhsMode == rhsMode
    case (.third(let lhsByName), .third(let rhsByName)):
      return lhsByName.byName == rhsByName.byName
    default:
      return false
    }
  }
}

class HybridRiveView: HybridRiveViewSpec {
  func play() throws -> NitroModules.Promise<Void> {
    return Promise.async {
      try await self.getRiveView().play()
    }
  }

  func pause() throws -> NitroModules.Promise<Void> {
    return Promise.async {
      try await self.getRiveView().pause()
    }
  }

  func reset() throws -> NitroModules.Promise<Void> {
    return Promise.async {
      try await self.getRiveView().reset()
    }
  }

  func playIfNeeded() {
    MainActor.assumeIsolated {
      try? self.getRiveView().playIfNeeded()
    }
  }

  // MARK: View Props
  var dataBind: HybridDataBindMode? {
    didSet {
      if !dataBind.isEqual(to: oldValue) {
        dataBindingChanged = true
      }
    }
  }

  var artboardName: String? { didSet { needsReload = true } }
  var stateMachineName: String? { didSet { needsReload = true } }
  var autoPlay: Bool? { didSet { needsReload = true } }
  var file: (any HybridRiveFileSpec) = HybridRiveFile() {
    didSet { needsReload = true }
  }
  var alignment: Alignment?
  var fit: Fit?
  var layoutScaleFactor: Double?
  var onError: (RiveError) -> Void = { _ in }

  func awaitViewReady() throws -> Promise<Bool> {
    return Promise.async { [self] in
      return try await getRiveView().awaitViewReady()
    }
  }

  func bindViewModelInstance(viewModelInstance: (any HybridViewModelInstanceSpec)) throws {
    guard let vmi = (viewModelInstance as? HybridViewModelInstance)?.viewModelInstance
    else { return }
    try MainActor.assumeIsolated {
      try getRiveView().bindViewModelInstance(viewModelInstance: vmi)
    }
  }

  func getViewModelInstance() throws -> (any HybridViewModelInstanceSpec)? {
    return try MainActor.assumeIsolated {
      guard let vmi = try getRiveView().getViewModelInstance() else { return nil }
      guard let hybridFile = file as? HybridRiveFile, let worker = hybridFile.worker else {
        throw RuntimeError.error(withMessage: "No worker available from file")
      }
      return HybridViewModelInstance(viewModelInstance: vmi, worker: worker)
    }
  }

  func onEventListener(onEvent: @escaping (UnifiedRiveEvent) -> Void) throws {
    throw RuntimeError.error(withMessage: "Events are not supported in the experimental iOS API")
  }

  func removeEventListeners() throws {
    throw RuntimeError.error(withMessage: "Events are not supported in the experimental iOS API")
  }

  func setNumberInputValue(name: String, value: Double, path: String?) throws {
    try MainActor.assumeIsolated {
      try getRiveView().setNumberInputValue(name: name, value: Float(value), path: path)
    }
  }

  func getNumberInputValue(name: String, path: String?) throws -> Double {
    return try MainActor.assumeIsolated {
      try Double(getRiveView().getNumberInputValue(name: name, path: path))
    }
  }

  func setBooleanInputValue(name: String, value: Bool, path: String?) throws {
    try MainActor.assumeIsolated {
      try getRiveView().setBooleanInputValue(name: name, value: value, path: path)
    }
  }

  func getBooleanInputValue(name: String, path: String?) throws -> Bool {
    return try MainActor.assumeIsolated {
      try getRiveView().getBooleanInputValue(name: name, path: path)
    }
  }

  func triggerInput(name: String, path: String?) throws {
    try MainActor.assumeIsolated {
      try getRiveView().triggerInput(name: name, path: path)
    }
  }

  func setTextRunValue(name: String, value: String, path: String?) throws {
    try MainActor.assumeIsolated {
      try getRiveView().setTextRunValue(name: name, value: value, path: path)
    }
  }

  func getTextRunValue(name: String, path: String?) throws -> String {
    return try MainActor.assumeIsolated {
      try getRiveView().getTextRunValue(name: name, path: path)
    }
  }

  // MARK: Views
  var view: UIView = RiveReactNativeView()
  func getRiveView() throws -> RiveReactNativeView {
    guard let riveView = view as? RiveReactNativeView else {
      throw RuntimeError.error(withMessage: "RiveReactNativeView is null or not configured")
    }
    return riveView
  }

  // MARK: Update
  func afterUpdate() {
    logged(tag: "HybridRiveView", note: "afterUpdate") {
      guard let hybridFile = file as? HybridRiveFile else {
        RCTLogError("[HybridRiveView] file is not HybridRiveFile: \(type(of: file))")
        return
      }
      guard let riveFile = hybridFile.file else {
        RCTLogError("[HybridRiveView] hybridFile.file is nil")
        return
      }

      let config = ExperimentalViewConfiguration(
        artboardName: artboardName,
        stateMachineName: stateMachineName,
        autoPlay: autoPlay ?? DefaultConfiguration.autoPlay,
        file: riveFile,
        fit: toExperimentalFit(fit, alignment: alignment, layoutScaleFactor: layoutScaleFactor),
        bindData: try dataBind.toExperimentalBindData()
      )

      try MainActor.assumeIsolated {
        let riveView = try getRiveView()
        riveView.configure(
          config, dataBindingChanged: dataBindingChanged, reload: needsReload,
          initialUpdate: initialUpdate)
        needsReload = false
        dataBindingChanged = false
        initialUpdate = false
      }
    }
  }

  // MARK: Internal State
  private var needsReload = false
  private var dataBindingChanged = false
  private var initialUpdate = true

  // MARK: Helpers
  private func toExperimentalFit(
    _ fit: Fit?,
    alignment: Alignment?,
    layoutScaleFactor: Double?
  ) -> RiveRuntime.Fit {
    let expAlignment = toExperimentalAlignment(alignment) ?? .center

    switch fit ?? .contain {
    case .fill: return .fill(alignment: expAlignment)
    case .contain: return .contain(alignment: expAlignment)
    case .cover: return .cover(alignment: expAlignment)
    case .fitwidth: return .fitWidth(alignment: expAlignment)
    case .fitheight: return .fitHeight(alignment: expAlignment)
    case .none: return .none(alignment: expAlignment)
    case .scaledown: return .scaleDown(alignment: expAlignment)
    case .layout:
      if let sf = layoutScaleFactor {
        return .layout(scaleFactor: .explicit(Float(sf)))
      }
      return .layout(scaleFactor: .automatic)
    }
  }

  private func toExperimentalAlignment(_ alignment: Alignment?) -> RiveRuntime.Alignment? {
    guard let alignment = alignment else { return nil }

    switch alignment {
    case .topleft: return .topLeft
    case .topcenter: return .topCenter
    case .topright: return .topRight
    case .centerleft: return .centerLeft
    case .center: return .center
    case .centerright: return .centerRight
    case .bottomleft: return .bottomLeft
    case .bottomcenter: return .bottomCenter
    case .bottomright: return .bottomRight
    }
  }
}

extension HybridRiveView {
  func logged(tag: String, note: String? = nil, _ fn: () throws -> Void) {
    do {
      return try fn()
    } catch let e {
      let noteString = note.map { " \($0)" } ?? ""
      let errorMessage = "[RIVE] \(tag)\(noteString) \(e.localizedDescription)"

      let riveError = RiveError(
        message: errorMessage,
        type: .unknown
      )
      onError(riveError)
    }
  }
}

File: android/src/new/java/com/margelo/nitro/rive/HybridRiveView.kt
package com.margelo.nitro.rive

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.core.Promise
import com.rive.BindData
import com.rive.RiveReactNativeView
import com.rive.ViewConfiguration
import app.rive.Fit as RiveFit
import app.rive.Alignment as RiveAlignment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

fun Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName?.toBindData(): BindData {
  if (this == null) return BindData.Auto

  return when (this) {
    is Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName.First -> {
      val instance = (this.asFirstOrNull() as? HybridViewModelInstance)?.viewModelInstance
        ?: throw IllegalStateException("Invalid ViewModelInstance")
      BindData.Instance(instance)
    }
    is Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName.Second -> {
      when (this.asSecondOrNull()) {
        DataBindMode.AUTO -> BindData.Auto
        DataBindMode.NONE -> BindData.None
        else -> BindData.None
      }
    }
    is Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName.Third -> {
      val name = this.asThirdOrNull()?.byName ?: throw IllegalStateException("Missing byName value")
      BindData.ByName(name)
    }
  }
}

object DefaultConfiguration {
  const val AUTOPLAY = true
  val ALIGNMENT = RiveAlignment.Center
  val LAYOUTSCALEFACTOR = null
}

@Keep
@DoNotStrip
class HybridRiveView(val context: ThemedReactContext) : HybridRiveViewSpec() {
  companion object {
    private const val TAG = "HybridRiveView"
  }

  //region Lifecycle
  override fun dispose() {
    // Eager teardown on JS unmount (RiveView calls dispose() in its unmount
    // effect, PR #202); RiveViewManager.onDropViewInstance is the backstop.
    // Nitro invokes this on the JS thread, but the view's teardown touches
    // main-thread state (Choreographer), so hop to main.
    val riveView = view
    UiThreadUtil.runOnUiThread {
      riveView.dispose()
    }
    super.dispose()
  }
  //endregion

  override val view: RiveReactNativeView = RiveReactNativeView(context).apply {
    onError = { msg ->
      this@HybridRiveView.onError(RiveError(type = RiveErrorType.UNKNOWN, message = msg))
    }
    onStop = {
      this@HybridRiveView.onStop()
    }
  }
  private var needsReload = false
  private var dataBindingChanged = false
  private var initialUpdate = true
  private var registeredFile: HybridRiveFile? = null

  override var artboardName: String? = null
    set(value) {
      changed(field, value) { field = it }
    }
  override var stateMachineName: String? = null
    set(value) {
      changed(field, value) { field = it }
    }
  override var autoPlay: Boolean? = null
    set(value) {
      changed(field, value) { field = it }
    }
  override var file: HybridRiveFileSpec = HybridRiveFile(null, HybridRiveFileFactory.getSharedWorker())
    set(value) {
      if (field != value) {
        registeredFile?.unregisterView(this)
        registeredFile = null
      }
      changed(field, value) { field = it }
    }
  override var alignment: Alignment? = null
  override var fit: Fit? = null
  override var layoutScaleFactor: Double? = null

  // The render loop can only skip frames, so a range is honored best-effort
  // as a cap at preferred ?? maximum.
  override var frameRate: Variant_Double_FrameRateRange? = null
    set(value) {
      field = value
      view.frameRate = value?.match(
        first = { it },
        second = { range -> range.preferred ?: range.maximum }
      )
    }

  // Accepted for API parity; semantics support is pending in the upstream
  // rive-android runtime (iOS-only for now).
  override var semantics: Semantics? = null
  override var dataBind: Variant_HybridViewModelInstanceSpec_DataBindMode_DataBindByName? = null
    set(value) {
      if (field != value) {
        field = value
        dataBindingChanged = true
      }
    }
  override var onError: (error: RiveError) -> Unit = {}
  override var onStop: () -> Unit = {}

  override fun awaitViewReady(): Promise<Boolean> {
    return Promise.async {
      withContext(Dispatchers.Main) {
        view.awaitViewReady()
      }
    }
  }

  override fun bindViewModelInstance(viewModelInstance: HybridViewModelInstanceSpec) =
    executeOnUiThread {
      val hybridVmi = viewModelInstance as? HybridViewModelInstance ?: return@executeOnUiThread
      view.bindViewModelInstance(hybridVmi.viewModelInstance)
    }

  override fun getViewModelInstance(): HybridViewModelInstanceSpec? {
    val vmi = view.getViewModelInstance() ?: return null
    val hybridFile = file as? HybridRiveFile ?: return null
    // The view owns (or JS-side dataBind owns) this instance — the wrapper
    // must not close it on dispose.
    return HybridViewModelInstance(vmi, hybridFile.riveWorker, hybridFile, ownsInstance = false)
  }

  override fun play(): Promise<Unit> = Promise.async { view.play() }
  override fun pause(): Promise<Unit> = Promise.async { view.pause() }
  override fun reset(): Promise<Unit> = Promise.async { view.reset() }
  override fun playIfNeeded() = view.playIfNeeded()

  override fun onEventListener(onEvent: (event: UnifiedRiveEvent) -> Unit) {
    throw UnsupportedOperationException("Events are not supported in the experimental Android API")
  }

  override fun removeEventListeners() {
    throw UnsupportedOperationException("Events are not supported in the experimental Android API")
  }

  override fun setNumberInputValue(name: String, value: Double, path: String?) {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  override fun getNumberInputValue(name: String, path: String?): Double {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  override fun setBooleanInputValue(name: String, value: Boolean, path: String?) {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  override fun getBooleanInputValue(name: String, path: String?): Boolean {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  override fun triggerInput(name: String, path: String?) {
    throw UnsupportedOperationException("SMI inputs not supported in experimental API")
  }

  override fun setTextRunValue(name: String, value: String, path: String?) {
    throw UnsupportedOperationException("Text runs not supported in experimental API")
  }

  override fun getTextRunValue(name: String, path: String?): String {
    throw UnsupportedOperationException("Text runs not supported in experimental API")
  }

  fun refreshAfterAssetChange() {
    afterUpdate()
  }

  override fun afterUpdate() {
    logged(TAG, "afterUpdate") {
      val hybridFile = file as? HybridRiveFile
      val riveFile = hybridFile?.riveFile ?: return@logged

      val convertedAlignment = convertAlignment(alignment) ?: DefaultConfiguration.ALIGNMENT
      val convertedFit = convertFit(fit, layoutScaleFactor?.toFloat(), convertedAlignment)
        ?: RiveFit.Contain(convertedAlignment)
      val config = ViewConfiguration(
        artboardName = artboardName,
        stateMachineName = stateMachineName,
        autoPlay = autoPlay ?: DefaultConfiguration.AUTOPLAY,
        riveFile = riveFile,
        riveWorker = HybridRiveFileFactory.getSharedWorker(),
        alignment = convertedAlignment,
        fit = convertedFit,
        layoutScaleFactor = layoutScaleFactor?.toFloat() ?: DefaultConfiguration.LAYOUTSCALEFACTOR,
        bindData = dataBind.toBindData()
      )
      view.configure(config, dataBindingChanged = dataBindingChanged, needsReload, initialUpdate = initialUpdate)

      if (needsReload && hybridFile != null) {
        hybridFile.registerView(this)
        registeredFile = hybridFile
      }

      needsReload = false
      dataBindingChanged = false
      initialUpdate = false
      super.afterUpdate()
    }
  }

  private fun <T> changed(current: T, new: T, setter: (T) -> Unit) {
    if (current != new) {
      setter(new)
      needsReload = true
    }
  }

  private fun executeOnUiThread(action: () -> Unit) {
    context.currentActivity?.runOnUiThread {
      try {
        action()
      } catch (e: Exception) {
        throw RuntimeException(e.message, e)
      }
    }
  }

  private fun convertAlignment(alignment: Alignment?): RiveAlignment? {
    if (alignment == null) return null
    return when (alignment) {
      Alignment.TOPLEFT -> RiveAlignment.TopLeft
      Alignment.TOPCENTER -> RiveAlignment.TopCenter
      Alignment.TOPRIGHT -> RiveAlignment.TopRight
      Alignment.CENTERLEFT -> RiveAlignment.CenterLeft
      Alignment.CENTER -> RiveAlignment.Center
      Alignment.CENTERRIGHT -> RiveAlignment.CenterRight
      Alignment.BOTTOMLEFT -> RiveAlignment.BottomLeft
      Alignment.BOTTOMCENTER -> RiveAlignment.BottomCenter
      Alignment.BOTTOMRIGHT -> RiveAlignment.BottomRight
    }
  }

  private fun convertFit(
    fit: Fit?,
    layoutScaleFactor: Float? = null,
    alignment: RiveAlignment = DefaultConfiguration.ALIGNMENT
  ): RiveFit? {
    if (fit == null) return null
    return when (fit) {
      // Fill and Layout cover the whole surface, so they carry no alignment.
      Fit.FILL -> RiveFit.Fill
      Fit.CONTAIN -> RiveFit.Contain(alignment)
      Fit.COVER -> RiveFit.Cover(alignment)
      Fit.FITWIDTH -> RiveFit.FitWidth(alignment)
      Fit.FITHEIGHT -> RiveFit.FitHeight(alignment)
      Fit.NONE -> RiveFit.None(alignment)
      Fit.SCALEDOWN -> RiveFit.ScaleDown(alignment)
      Fit.LAYOUT -> RiveFit.Layout(scaleFactor = layoutScaleFactor ?: context.resources.displayMetrics.density)
    }
  }

  fun logged(tag: String, note: String? = null, fn: () -> Unit) {
    try {
      fn()
    } catch (e: Exception) {
      val message = e.message ?: e.toString()
      val noteString = note?.let { " $it" } ?: ""
      val errorMessage = "[RIVE] $tag$noteString $message"
      val riveError = RiveError(
        type = RiveErrorType.UNKNOWN,
        message = errorMessage
      )
      onError(riveError)
    }
  }
}
